File: SMSdisfwais/SMSdisfwais.m
clear all
close all
clc

%% Fisher's LDA for WAIS data

load('wais.dat')

mdl        = ClassificationDiscriminant.fit(wais(:,3:end),wais(:,1),'Prior','uniform');
% LDA model prediction
pred       = predict(mdl,wais(:,3:end));
% truth-prediction 
label      = wais(:,1);
[C,order]  = confusionmat(label,pred);

cols       = cellstr(strvcat('         ','         ','Prediction','         '));
rows       = cellstr(strvcat('         ','Truth     ','         '));
rows1      = cellstr(strvcat('         ',order));
cols1      = cellstr(strvcat(order(1,:), num2str(C(:,1))));
cols2      = cellstr(strvcat(order(2,:), num2str(C(:,2))));
table      = [rows, rows1, cols1, cols2];
table      = [cols'; table];
disp(table)

%% compute apparent error rate (APER)
aper       = (sum(sum(C))-sum(diag(C)))/sum(sum(C));
disp(aper)

%% compute actual error rate (AER)
correct=0;
for i = 1:size(wais,1)
    subset      = wais;
    subset(i,:) = [];
    mdl         = ClassificationDiscriminant.fit(subset(:,3:end),subset(:,1),'Prior','uniform');
    pred        = predict(mdl,wais(i,3:end));
    if (pred==wais(i,1))
        correct=correct+1;
    end
end


aer  = 1-correct/size(wais,1);
disp(aer)

File: SMSdisfwais/SMSdisfwais.R
# Clear workspace
rm(list=ls(all=TRUE))
graphics.off()

# install and load packages
libraries = c("stats","MASS")
lapply(libraries, function(x) if (!(x %in% installed.packages())) {
    install.packages(x)})
lapply(libraries, library, quietly = TRUE, character.only = TRUE)

# setwd("C:/...") # please change the working directory
load("wais.rda")

fisher.w=lda(group~.-subject,prior=c(0.5,0.5),data=wais)
prediction=predict(fisher.w, wais)$class

t=table(wais$group,prediction)

print(t)
aper=(sum(t)-sum(diag(t)))/sum(t)
print(aper)

correct=0

for (i in 1:nrow(wais)) {
  fisher.t=lda(group~.-subject,prior=c(0.5,0.5),subset=-i,data=wais)
  predict=predict(fisher.t, wais[i,])$class
  if (predict==wais[i,"group"]) correct=correct+1
}

aer=1-correct/nrow(wais)
print(aer)
